addons/udes_stock/tests/common.py

```diff
@@ -0,0 +1,467 @@
+            * all picking types as a copy of the default with TEST_ prepended
+            * a simple inbound route goods_in->putaway
+            * a simple outbound route pick->goods_out
+        """
+        cls._setup_picking_types()
+        cls._setup_locations()
+        cls._setup_routes()
+
+    @classmethod
+    def _setup_picking_types(cls):
+        """ Additional test picking types - prepended with TEST """
+        cls._set_test_picking_type('in', 'goods_in')
+        cls._set_test_picking_type('pick')
+        # this order is important as int is overwritten
+        cls._set_test_picking_type('int', 'putaway')
+        cls._set_test_picking_type('int', 'internal')
+        cls._set_test_picking_type('out', 'goods_out')
+        cls._set_test_picking_type('out', 'trailer_dispatch')
+
+    @classmethod
+    def _set_test_picking_type(cls, ptype, name=None):
+        # TODO: Tidy this bit up
+        Warehouse = cls.env['stock.warehouse']
+        user_warehouse = Warehouse.search([('name', '=', 'Warehouse')])
+        if name is None:
+            name = ptype
+
+        copy_vals = {'name': 'TEST_{}'.format(name.upper()), 'active': True}
+        wh_attr = getattr(user_warehouse, '{}_type_id'.format(ptype))
+        new_pt = wh_attr.copy(copy_vals)
+        setattr(cls, 'picking_type_{}'.format(name), new_pt)
+        setattr(user_warehouse, '{}_type_id'.format(ptype), new_pt)
+
+    @classmethod
+    def create_product(cls, name, **kwargs):
+        """ Create and return a product """
+        Product = cls.env['product.product']
+        vals = {
+            'name': 'Test product {}'.format(name),
+            'barcode': 'product{}'.format(name),
+            'default_code': 'product{}'.format(name),
+            'type': 'product',
+        }
+        vals.update(kwargs)
+        return Product.create(vals)
+
+    @classmethod
+    def create_quant(cls, product_id, location_id, qty, serial_number=None, **kwargs):
+        """ Create and return a quant of a product at location """
+        Quant = cls.env['stock.quant']
+        vals = {
+            'product_id': product_id,
+            'location_id': location_id,
+            'quantity': qty,
+        }
+        if serial_number:
+            lot = cls.create_lot(product_id, serial_number)
+            vals['lot_id'] = lot.id
+        vals.update(kwargs)
+        return Quant.create(vals)
+
+    @classmethod
+    def create_lot(cls, product_id, serial_number, **kwargs):
+        Lot = cls.env['stock.production.lot']
+
+        vals = {
+            'name': serial_number,
+            'product_id': product_id,
+        }
+        vals.update(kwargs)
+        return Lot.create(vals)
+
+    @classmethod
+    def create_package(self, **kwargs):
+        """ Create and return a new package """
+        Package = self.env['stock.quant.package']
+        vals = {}
+        vals.update(kwargs)
+        return Package.create(vals)
+
+    @classmethod
+    def _setup_locations(cls):
```

Comment:
perhaps some class variables might allow better control over these setup functions

Reply:
Also most of these allow for multi creates which would speed up test setup :)

Reply:
Can you put an example of what type of control?

Reply:
I can't remember what I was getting at here :/